Post Editor: Add stats for the Gutenberg opt-in dialog.

kwight opened the pull request:
This PR adds stats for the buttons in the opt-in dialog.

<img width="1100" alt="screen shot 2018-09-11 at 4 32 06 pm" src="https://user-images.githubusercontent.com/349751/45393345-58af8600-b5e0-11e8-94e1-b1c912732fb0.png">

**Testing**

* Observe stats activity in the console with `localStorage.setItem( 'debug', 'calypso:analytics*' );`
* Open the post editor to see the opt-in visual at the bottom of the sidebar.
* Clicking the dialog's `X` button and clicking outside of the dialog close the dialog, with no stats recorded.
* Click both the "Try..." and "stick to classic" buttons.
* "Try...": closes the dialog and forwards the user to a `/gutenberg` route.
* "classic" just closes the dialog.
* Both buttons send the appropriate stats, eg:

<img width="983" alt="screen shot 2018-09-13 at 2 41 59 pm" src="https://user-images.githubusercontent.com/349751/45517622-9be63200-b763-11e8-8ce5-38f49e032c47.png">

alisterscott commented:
Apologies I didn't see WIP in title and added the review label - sorry

dmsnell commented:
Can we consider doing this in the Analytics middleware to watch for the opt-in or opt-out actions and then record the metrics when they fly by instead of wiring it into the component itself?

There are two ways we can do this: colocate the analytics meta inside the component's module; create a separate external watcher for the action by injecting the middleware.

The easiest for this PR could be simply to move those wrapped functions into `connect()` and `mapDispatchToProps` so we can preserve the pure component and the default behavior but without needing to do much more work.

kwight commented:
OK, I think I've got things running properly; I've taken the liberty of adding a `GUTENBERG_OPT_IN` action and creators so that I can wrap these with the stats (we'll need them soon enough anyways).

One thing I'm not sure of is if the `href` handling of the opt-in button gets in the way of calling the `optInToGutenberg` creator; without the href, I can see the `GUTENBERG_OPT_IN` action happen, but with it, I think it's missed?

Copons commented on client/post-editor/editor-gutenberg-opt-in-dialog/index.jsx:
```diff
@@ -14,39 +14,66 @@ import Gridicon from 'gridicons';
+		const { translate, gutenbergURL, isDialogVisible, hideDialog } = this.props;
 		const buttons = [
-			<Button key="gutenberg" href={ gutenbergURL } primary>
+			<Button key="gutenberg" href={ gutenbergURL } onClick={ this.optInAndHideDialog } primary>
```
Regardless of this potential race condition, `GUTENBERG_OPT_IN` and `GUTENBERG_OPT_IN_DIALOG_IS_SHOWING` are still dispatched correctly upon entering Gutenberg.

Though, we might improve this by dispatching a `navigate` action in the `onClick` instead of relying on the `href`.

```es6
import { navigate } from 'state/ui/actions';

// ...

optInAndOpenGutenberg = () => {
  this.props.optIn();
  this.props.hideDialog(); // needed because otherwise it will open again on next load 🤔
  this.props.goToGutenberg( this.props.gutenbergURL );
}

// ...

goToGutenberg: gutenbergURL => dispatch( navigate( gutenbergURL ) );
```

Note that we could also streamline the `goToGutenberg` dispatch by providing the URL to it via `mergeProps` (probably overkill tho 🤷‍♂️ )

kwight replied:
>  It feels like we can dispatch one, and move the side-effect to middleware (either by setting middleware or dispatching another action)

Maybe we can leave this until a later PR that adds a proper `GUTENBERG_OPT_IN` action.

gwwar commented:
Can we pare down scope a bit? I think for this pass all we want to know is when folks click on the try it button or dismiss the modal. I think it's a little early for that opt-in redux action.

kwight commented on client/post-editor/editor-gutenberg-opt-in-dialog/index.jsx:
```diff
@@ -18,23 +18,40 @@ import { hideGutenbergOptInDialog } from 'state/ui/gutenberg-opt-in-dialog/actio
+		optOut: PropTypes.func,
+	};
+
+	onCloseDialog = () => {
```
Noting that this just moves `onCloseDialog` from under `render` to above it.

kwight commented:
> Can we pare down scope a bit? I think for this pass all we want to know is when folks click on the try it button or dismiss the modal. I think it's a little early for that opt-in redux action.

I've removed the `GUTENBERG_OPT_IN` action creators, and just added `optIn` and `outOut` props through `mapDispatchToProps`, which handle the stats and dialog closing – this should be much simpler to move forward with 👍
